In [1]:
# antes de empezar importamos la librería de NumPy para poder trabajar con todos sus métodos. 
import numpy as np

In [2]:
# definimos un array bidimensional
# en este caso estamos generando un array de 5 filas y 4 columnas con números aleatorios entre el 0 y el 1

bidimensional = np.random.rand(5, 4)
print(f"El array bidimensional que tenemos es: \n {bidimensional}")

# desgranemos la indexación en este tipo de arrays
# si accedemos al elemento 0 del array obtenemos a la fila que queramos de nuestro array

print(f"\nEl primer elemento de nuestro array es: \n {bidimensional[0]}")

El array bidimensional que tenemos es: 
 [[0.78415423 0.69118324 0.59972248 0.98798808]
 [0.58703524 0.87943394 0.89662338 0.1073667 ]
 [0.79053033 0.20318369 0.20645637 0.5421861 ]
 [0.60561998 0.75280002 0.59636169 0.67319644]
 [0.445758   0.42933543 0.76492663 0.77372507]]

El primer elemento de nuestro array es: 
 [0.78415423 0.69118324 0.59972248 0.98798808]


In [3]:
# en este caso reutilizaremos el array bidimensional que creamos en el apartado anterior
print(f"El array bidimensional con el que vamos a trabajar es: \n {bidimensional}")

# creamos la máscara, en este caso nuestra condición será: ¿qué elementos son mayores que 0.5?
print("-----------------------------")
print(f"Primer paso: creación de una máscara booleana \n {bidimensional > 0.5}")

# aplicamos la máscara creada en el paso anterior, en este caso, NumPy nos devolverá solo los elementos 
# que coinciden con la condicion
print("-----------------------------")
print(f"Segundo paso: aplicar la máscara booleana \n {bidimensional[bidimensional > 0.5]}")

El array bidimensional con el que vamos a trabajar es: 
 [[0.41584614 0.82312683 0.19433902 0.81775218]
 [0.76281141 0.86328659 0.94662046 0.87703294]
 [0.02764513 0.03467823 0.14428307 0.98926624]
 [0.51561502 0.17237538 0.59164942 0.77651838]
 [0.76409676 0.31078056 0.00390948 0.17254908]]
-----------------------------
Primer paso: creación de una máscara booleana 
 [[False  True False  True]
 [ True  True  True  True]
 [False False False  True]
 [ True False  True  True]
 [ True False False False]]
-----------------------------
Segundo paso: aplicar la máscara booleana 
 [0.82312683 0.81775218 0.76281141 0.86328659 0.94662046 0.87703294
 0.98926624 0.51561502 0.59164942 0.77651838 0.76409676]


Imaginemos ahora que queremos seleccionar ciertos valores basados en dos condiciones. En este caso:

- Usaremos el operador `&` para indicar que queremos que se cumplan las dos condiciones. Lo que conocíamos hasta ahora como `and`.


- Usaremos el operador `|` para indicar que se cumpla una condición y otra. Lo que conocíamos hasta ahora como `or`. 

In [7]:
# en este caso reutilizaremos el array bidimensional que creamos en el apartado anterior
print(f"El array bidimensional con el que vamos a trabajar es: \n {bidimensional}")


# queremos seleccionar aquellos números que sean MENORES que 0.2 
# o MAYORES que 0.7. FIJATE QUE CADA CONDICIÓN VA ENTRE PARÉNTESIS SEPARADA POR EL OPERADOR "|"

print("-----------------------------")
resultado_filtradoI = bidimensional[(bidimensional < 0.2) |  (bidimensional >0.7)]
print(f"El resultado del filtrado < 0.2 o > 0.7 es: \n {resultado_filtradoI}")

El array bidimensional con el que vamos a trabajar es: 
 [[0.41584614 0.82312683 0.19433902 0.81775218]
 [0.76281141 0.86328659 0.94662046 0.87703294]
 [0.02764513 0.03467823 0.14428307 0.98926624]
 [0.51561502 0.17237538 0.59164942 0.77651838]
 [0.76409676 0.31078056 0.00390948 0.17254908]]
-----------------------------
El resultado del filtrado < 0.2 o > 0.7 es: 
 [0.82312683 0.19433902 0.81775218 0.76281141 0.86328659 0.94662046
 0.87703294 0.02764513 0.03467823 0.14428307 0.98926624 0.17237538
 0.77651838 0.76409676 0.00390948 0.17254908]


## Filtrado con `np.where()`

Permite realizar una evaluación condicional sobre un *array*. La función `np.where()` devuelve un nuevo array con los elementos seleccionados según la condición especificada.

Su sintaxis básica es:
```python
np.where(condición, valor_si_verdadero, valor_si_falso)
```

Donde:

- **condición**: Es una expresión booleana que define la condición a evaluar. 

- **valor_si_verdadero** (opcional):  Es el valor o array que se selecciona si la condición es verdadera.

- **valor_si_falso** (opcional): Es el valor o array que se selecciona si la condición es falsa.

In [10]:
print(f"El array con el que vamos a trabajar en este apartado de la lección es: \n {bidimensional}")

# usamos el método 'np.where()' para evaluar una condición, en este caso para encontrar los valores mayores que 0.8

resultado_where = np.where(bidimensional > 0.8)
print("--------------------------")
print("El resultado del np.where() es: \n", resultado_where)

El array con el que vamos a trabajar en este apartado de la lección es: 
 [[0.41584614 0.82312683 0.19433902 0.81775218]
 [0.76281141 0.86328659 0.94662046 0.87703294]
 [0.02764513 0.03467823 0.14428307 0.98926624]
 [0.51561502 0.17237538 0.59164942 0.77651838]
 [0.76409676 0.31078056 0.00390948 0.17254908]]
--------------------------
El resultado del np.where() es: 
 (array([0, 0, 1, 1, 1, 2], dtype=int64), array([1, 3, 1, 2, 3, 3], dtype=int64))


Seguramente hayan notado que la salida del resulado, pasemos a explicarlo un poco!

tengamos en cuenta nuestro siguiente array y el resultado de nuestro where:

```python
[[0.41584614 0.82312683 0.19433902 0.81775218]
 [0.76281141 0.86328659 0.94662046 0.87703294]
 [0.02764513 0.03467823 0.14428307 0.98926624]
 [0.51561502 0.17237538 0.59164942 0.77651838]
 [0.76409676 0.31078056 0.00390948 0.17254908]]

 El resultado del np.where() es: 
 (array([0, 0, 1, 1, 1, 2], dtype=int64), array([1, 3, 1, 2, 3, 3], dtype=int64))
```

nuestro where ha encontrado las posiciones de los elementos en array que son mayores que 0.8. El resultado es una tupla de dos arrays:

*   El primer array contiene los índices de las filas.
*   El segundo array contiene los índices de las columnas.


El detalle es el siguiente:

*   array([0, 0, 1, 1, 1, 2], dtype=int64) representa las filas.
*   array([1, 3, 1, 2, 3], dtype=int64) representa las columnas.

Al combinar cada fila y columna con su posicion, nos da una ubicacion de los elementos que cumplen mi condicion. por ejemplo el primer valor del array de filas es 0 y el primer valor de array de columnas es 1, por lo tanto el elemento que cumple la condicion esta en [0,1]

Vamos a verificar los valores en esas posiciones para confirmar que son mayores que 0.8:

*   array[0, 1] -> 0.82312683
*   array[0, 3] -> 0.81775218
*   array[1, 1] -> 0.86328659
*   array[1, 2] -> 0.94662046
*   array[1, 3] -> 0.87703294
*   array[2, 3] -> 0.98926624


Efectivamente, todos estos valores son mayores que 0.8!

Ok, dijimos anteriormente que nosotros utilizando where, podriamos asignar valores segun se cumpla o no una condicion, para este caso asignaremos xxx cuando se cumpla la condicion (es decir, la condicion sea verdadera) y asignaremos ooo cuando no lo sea.

In [9]:
print(f"El array con el que vamos a trabajar en este apartado de la lección es: \n {bidimensional}")

# usamos el método 'np.where()' para evaluar una condición, en este caso para encontrar los valores mayores que 0.8

resultado_where = np.where(bidimensional > 0.8,"xxx","ooo")
print("--------------------------")
print("El resultado del np.where() es: \n", resultado_where)

El array con el que vamos a trabajar en este apartado de la lección es: 
 [[0.41584614 0.82312683 0.19433902 0.81775218]
 [0.76281141 0.86328659 0.94662046 0.87703294]
 [0.02764513 0.03467823 0.14428307 0.98926624]
 [0.51561502 0.17237538 0.59164942 0.77651838]
 [0.76409676 0.31078056 0.00390948 0.17254908]]
--------------------------
El resultado del np.where() es: 
 [['ooo' 'xxx' 'ooo' 'xxx']
 ['ooo' 'xxx' 'xxx' 'xxx']
 ['ooo' 'ooo' 'ooo' 'xxx']
 ['ooo' 'ooo' 'ooo' 'ooo']
 ['ooo' 'ooo' 'ooo' 'ooo']]


NOTA IMPORTANTE! En la demostracion anterior asignamos un tipo de dato str a nuestros valores evaluados de la matriz, esto podria ser un problema si necesitamos que los valores siempre sean numericos.

# Operaciones aritméticas con *arrays* 

*NumPy* proporciona una amplia gama de funciones y operaciones para realizar cálculos aritméticos en *arrays* de manera eficiente.

Algunas de las operaciones aritméticas comunes que puedes realizar con arrays de NumPy:

1. Suma: Usaremos el método `np.add()`
  
2. Resta: En este caso`np.subtract()`

3. Multiplicación: Usaremos `np.multiply()`

4. División: Método `np.divide()`

5. Potencia: Usaremos el método `np.power()`

6. Operaciones con escalares

Los operadores `+`, `-`, `*` y `/` también funcionan para realizar las operaciones correspondientes directamente entre *arrays* de *NumPy*. Sin embargo, aunque los podemos usar, los métodos aprendidos específicos de NumPy son los más adecuados.

In [11]:
# lo primero que vamos a hacer es definir dos arrays con los que trabajaremos. 
# en este caso vamos a generar dos arrays tridimensionales con números enteros entre el 1 y 10
# el array consta de 2 matrices, 2 filas y tres colummas 

array1 = np.random.randint(1,10, (2, 2, 3))
array2 = np.random.randint(1,10, (2, 2, 3))

print("El array 1 es:\n ", array1, '\n')
print("El array 2 es:\n ", array2)

# sumar dos arrays

print("--------------------------------")
suma = np.add(array1, array2)  # Suma los elementos correspondientes de los dos arrays
print("El resultado de la suma de los dos arrays es:\n", suma)


El array 1 es:
  [[[4 2 8]
  [9 7 3]]

 [[3 7 8]
  [9 5 1]]] 

El array 2 es:
  [[[3 5 8]
  [2 1 2]]

 [[1 8 3]
  [9 5 1]]]
--------------------------------
El resultado de la suma de los dos arrays es:
 [[[ 7  7 16]
  [11  8  5]]

 [[ 4 15 11]
  [18 10  2]]]


Como pueden ver, las operaciones se realizan por posicion, es decir, en este caso se suma la posicion [0][0][0] del primer array tridimensional con la posicion similar del segundo array tridimensional.

In [12]:
# misma logica en el caso de la resta:
print("El array 1 es:\n ", array1, '\n')
print("El array 2 es:\n ", array2)

# restar dos arrays
print("--------------------------------")
resta = np.subtract(array1, array2)  # Resta los elementos correspondientes de los dos arrays
print("El resultado de la resta de los dos arrays es:\n", resta)

El array 1 es:
  [[[4 2 8]
  [9 7 3]]

 [[3 7 8]
  [9 5 1]]] 

El array 2 es:
  [[[3 5 8]
  [2 1 2]]

 [[1 8 3]
  [9 5 1]]]
--------------------------------
El resultado de la resta de los dos arrays es:
 [[[ 1 -3  0]
  [ 7  6  1]]

 [[ 2 -1  5]
  [ 0  0  0]]]


In [13]:
# la multiplicacion mantiene la misma logica de operar por cada posicion entre los diferentes arrays.

print("El array 1 es:\n ", array1, '\n')
print("El array 2 es:\n ", array2)

# multiplicar dos arrays

print("--------------------------------")
multipicacion = np.multiply(array1, array2)  # Multiplicacion los elementos correspondientes de los dos arrays
print("El resultado de la multiplicación de los dos arrays es:\n", multipicacion)

El array 1 es:
  [[[4 2 8]
  [9 7 3]]

 [[3 7 8]
  [9 5 1]]] 

El array 2 es:
  [[[3 5 8]
  [2 1 2]]

 [[1 8 3]
  [9 5 1]]]
--------------------------------
El resultado de la multiplicación de los dos arrays es:
 [[[12 10 64]
  [18  7  6]]

 [[ 3 56 24]
  [81 25  1]]]


In [15]:
# Para la division, seguimos con la logica de operar por cada posicion entre los arrays.

print("El array 1 es:\n ", array1, '\n')
print("El array 2 es:\n ", array2)

# dividir dos arrays

print("--------------------------------")
division = np.divide(array1, array2)  # Division los elementos correspondientes de los dos arrays
print("El resultado de la división de los dos arrays es:\n", division)

El array 1 es:
  [[[3 4 5]
  [2 8 1]]

 [[6 2 1]
  [7 8 3]]] 

El array 2 es:
  [[[5 7 5]
  [2 5 1]]

 [[3 9 7]
  [6 6 3]]]
--------------------------------
El resultado de la división de los dos arrays es:
 [[[0.6        0.57142857 1.        ]
  [1.         1.6        1.        ]]

 [[2.         0.22222222 0.14285714]
  [1.16666667 1.33333333 1.        ]]]


In [14]:
#en este caso la potenciacion sigue las mismas reglas anteriores.

print("El array 1 es:\n ", array1, '\n')
print("El array 2 es:\n ", array2)

# elevar dos arrays

print("--------------------------------")
elevado = np.power(array1, array2)  # Elevar los elementos correspondientes de los dos arrays
print("El resultado de la potencia de los dos arrays es:\n", elevado)

El array 1 es:
  [[[4 2 8]
  [9 7 3]]

 [[3 7 8]
  [9 5 1]]] 

El array 2 es:
  [[[3 5 8]
  [2 1 2]]

 [[1 8 3]
  [9 5 1]]]
--------------------------------
El resultado de la potencia de los dos arrays es:
 [[[       64        32  16777216]
  [       81         7         9]]

 [[        3   5764801       512]
  [387420489      3125         1]]]


In [19]:
# operaciones con escalares
# definimos un escalar

escalar = 2

# y seguimos con nuestro array de ejemplo

print("El array 1 es:\n ", array1, '\n')

print("--------------------------------")

multiplicacion_escalar = array1 * escalar
print("El resultado de la multiplicación del array con el escalar es:\n", multiplicacion_escalar)

# Aca lo que sucede es que ese escalar se multiplicara por cada posicion de mi array definido.

El array 1 es:
  [[[3 4 5]
  [2 8 1]]

 [[6 2 1]
  [7 8 3]]] 

--------------------------------
El resultado de la multiplicación del array con el escalar es:
 [[[ 6  8 10]
  [ 4 16  2]]

 [[12  4  2]
  [14 16  6]]]
